cpuid/src/transformer/common.rs

```diff
@@ -64,14 +68,82 @@ pub fn update_brand_string_entry(
     Ok(())
 }
 
+pub fn update_cache_parameters_entry(
+    entry: &mut kvm_cpuid_entry2,
+    vm_spec: &VmSpec,
+) -> Result<(), Error> {
+    use cpu_leaf::leaf_cache_parameters::*;
+
+    match entry.eax.read_bits_in_range(&eax::CACHE_LEVEL_BITRANGE) {
+        // L1 & L2 Cache
+        1 | 2 => {
+            // The L1 & L2 cache is shared by at most 2 hyperthreads
+            entry.eax.write_bits_in_range(
+                &eax::MAX_CPUS_PER_CORE_BITRANGE,
+                (vm_spec.cpu_count > 1 && vm_spec.ht_enabled) as u32,
+            );
+        }
+        // L3 Cache
+        3 => {
+            // The L3 cache is shared among all the logical threads
+            entry.eax.write_bits_in_range(
+                &eax::MAX_CPUS_PER_CORE_BITRANGE,
+                u32::from(vm_spec.cpu_count - 1),
+            );
+        }
+        _ => (),
+    }
+
+    Ok(())
+}
+
+/// Replaces the `cpuid` entries corresponding to `function` with the entries from the host's cpuid.
+///
+pub fn use_host_cpuid_function(cpuid: &mut CpuId, function: u32) -> Result<(), Error> {
+    // copy all the CpuId entries, except for the ones with the provided function
+    let mut entries: Vec<kvm_cpuid_entry2> = Vec::new();
+    for entry in cpuid.mut_entries_slice().iter() {
+        if entry.function != function {
+            entries.push(*entry);
+        }
+    }
+
+    // add all the host leaves with the provided function
+    let mut count: u32 = 0;
+    while let Ok(entry) = get_cpuid(function, count) {
+        // check if there's enough space to add a new entry to the cpuid
+        if entries.len() == kvm::MAX_KVM_CPUID_ENTRIES {
+            return Err(Error::SizeLimitExceeded);
+        }
+
+        entries.push(kvm_cpuid_entry2 {
+            function,
+            index: count,
+            flags: 0,
+            eax: entry.eax,
+            ebx: entry.ebx,
+            ecx: entry.ecx,
+            edx: entry.edx,
+            padding: [0, 0, 0],
+        });
+        count += 1;
+    }
+
+    let cpuid2 = CpuId::from_entries(&entries);
+    *cpuid = cpuid2;
```

Comment:
This function completely replaces the `cpuid` struct with a new copy, which is not something the caller might expect. I.e. if, in the future, we decide to extend the `CpuId` struct to hold more data, that data would be lost upon returning from this function.